### PR TITLE
Add Async Cluster Support

### PR DESCRIPTION
**Pull request by jaymell**
Bring in code from https://github.com/redis-rs/redis-cluster-async
with minimal changes in order to allow for easier maintenance of
both projects and to better promote code reuse.

### CODE REVIEW

**Comment by jaymell** [comment]
@djc Wanted to get your feedback on the approach here. Still a work in progress, but I'm thinking it's the best way at this point to move forward with getting native async clustered support in redis-rs.

**Comment by djc** [comment]
Seems reasonable.

**Comment by utkarshgupta137** [comment]
The mpsc used in this implementation is unnecessary and adds an extra level of indirection. I ran some basic benchmarks and found that sending the message directly to the underlying pipeline in MultiplexedConnection is faster.

File: redis/src/cluster_async/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1239 @@
...
+}
+
+// Get slot data from connection.
+async fn get_slots<C>(addr: &str, connection: &mut C, use_tls: bool) -> RedisResult<Vec<Slot>>
+where
+    C: ConnectionLike,
+{
+    trace!("get_slots");
+    let mut cmd = Cmd::new();
+    cmd.arg("CLUSTER").arg("SLOTS");
```
Comment (shachlanAmazon) [inline redis/src/cluster_async/mod.rs:1100]: should use `CLUSTER SHARDS` on redis 7 - `CLUSTER SLOTS` is deprecated.
https://redis.io/commands/cluster-slots/

File: redis/src/cluster_async/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1239 @@
...
+            Connection(tx)
+        })
+    }
+}
+
+type SlotMap = BTreeMap<u16, String>;
+type ConnectionFuture<C> = future::Shared<BoxFuture<'static, C>>;
+type ConnectionMap<C> = HashMap<String, ConnectionFuture<C>>;
+
+struct Pipeline<C> {
```
Comment (shachlanAmazon) [inline redis/src/cluster_async/mod.rs:170]: I know this maintains the naming from the `aio` module, but `Pipeline` has two meanings here - both this and `CmdArg::Pipeline` (and `crate::Pipeline`). IMO this duality should be avoided, and the struct renamed to something like `ClusterConnection`, or just `Connection` and rename `Connection` to `NodeConnection`.

File: redis/src/cluster_async/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1239 @@
...
+    }
+
+    fn try_request(
+        &mut self,
+        info: &RequestInfo<C>,
+    ) -> impl Future<Output = (String, RedisResult<Response>)> {
+        // TODO remove clone by changing the ConnectionLike trait
+        let cmd = info.cmd.clone();
+        let (addr, conn) = if info.excludes.len() > 0 || info.slot.is_none() {
+            get_random_connection(&self.connections, Some(&info.excludes))
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:654]: Hey, I think your request here is a reasonable one, but this specific PR is probably not the place to do it. Do you mind creating a separate issue for this?

**Comment by jaymell** [comment]
@djc Hoping you can start reviewing this whenever you get a chance. I'd like to hold off on merging until after the next release is cut, but otherwise I think it's almost in a merge-able (which is not to say releasable) state.

I've tried to keep the commits as straightforward as possible but can certainly break things up further if helpful.

**Comment by djc** [comment]
Want to mark it as ready for review or would you prefer to hold off until after the next release?

**Comment by jaymell** [comment]
Marked it ready, would still prefer to hold off on merging until next release is created though. Also, I don't intend to release this as-is, I have a list of items [here](https://github.com/redis-rs/redis-rs/issues/745) I would to add in subsequent PRs before a release happens.

**Comment by trungda** [comment]
Thanks for doing this! redis-cluster-async definitely needs more support. What's the plan with open PRs in the redis-cluster-async? I also have one improvement (refresh-slot command doesn't handle username/password very well) that I'd like to open PR. What should I do? Should I still open PR against redis-cluster-async? Thanks.

**Comment by 0xWOF** [comment]
Just for curiosity. Why `cluster async connection` is provided by `cluster_async`  module instead `cluster_aio.rs` file? I think it is bit weired considering naming consistency. Because `normal async connection` is provided by `aio.rs` file and `cluster sync connection` is provided by `cluster.rs` file.

File: redis/src/cluster_async/mod.rs (hunk 2 above)
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:176]: Agreed that the naming is very confusing. Planning to address the confusion in subsequent PRs.

**Comment by Soufian20** [comment]
Is it also possible to adapt the redis::aio::ConnectionManager to also support the cluster_async::client?

-> https://docs.rs/redis/latest/redis/aio/struct.ConnectionManager.html

**Comment by jaymell** [comment]
> Is it also possible to adapt the redis::aio::ConnectionManager to also support the cluster_async::client?
> 
> -> https://docs.rs/redis/latest/redis/aio/struct.ConnectionManager.html

The clustered client itself should more or less perform the same functions as the connection manager by handling reconnections and retries when necessary. Is there something else I'm missing about `ConnectionManager` (which I personally have never worked with so am not terribly familiar)

**Comment by Soufian20** [comment]
> The clustered client itself should more or less perform the same functions as the connection manager by handling reconnections and retries when necessary. Is there something else I'm missing about `ConnectionManager` (which I personally have never worked with so am not terribly familiar)?

You're right it has more or less the same functionality. It just wraps the client and the connections as a proxy and try also an initial connection. I can see someone created already an issue for that -> https://github.com/redis-rs/redis-rs/issues/658

**Comment by Soufian20** [comment]
i tested your branch with this minimial cluster setup -> https://github.com/Soufian20/simple-redis-async-cluster
It works :+1:

File: redis/src/cluster_async/mod.rs [hunk 4]
```diff
@@ -0,0 +1,1239 @@
+//! This is a rust implementation for Redis cluster library.
+//!
+//! This library extends redis-rs library to be able to use cluster.
+//! Client impletemts traits of ConnectionLike and Commands.
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:4]: Will be addressed in future PR prior to next release.

File: redis/src/cluster_async/mod.rs (hunk 1 above)
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:1104]: Thanks, raised as separate issue.

File: redis/src/cluster_async/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1239 @@
...
+            if let ConnectionAddr::Unix(_) = info.addr {
+                return Err(RedisError::from((ErrorKind::InvalidClientConfig,
+                                             "This library cannot use unix socket because Redis's cluster command returns only cluster's IP and port.")));
+            }
+            nodes.push(info);
+        }
+
+        Ok(Client {
+            initial_nodes: nodes,
+            retries: Some(DEFAULT_RETRIES),
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:116]: Yes, will be addressed when clients are consolidated, in upcoming PR.

**Pull request merged**